org.librarysimplified.r2.views/src/main/java/org/librarysimplified/r2/views/SR2ReaderFragment.kt

```diff
@@ -101,57 +107,70 @@ class SR2ReaderFragment : Fragment() {
     FluentFuture.from(controllerFuture)
       .transform(
         Function<SR2ControllerType, Unit> { controller ->
-          this.onBookOpenSucceeded(controller!!)
+          onBookOpenSucceeded(controller!!)
         },
         MoreExecutors.directExecutor()
       )
       .catching(
         Throwable::class.java,
-        Function<Throwable, Unit> { exception ->
-          this.onBookOpenFailed(exception!!)
+        Function<Throwable, Unit> { e ->
+          onBookOpenFailed(e!!)
         }, MoreExecutors.directExecutor()
       )
   }
 
-  private fun onBookOpenSucceeded(controller: SR2ControllerType) {
-    this.logger.debug("onBookOpenSucceeded")
+  override fun onStop() {
+    super.onStop()
+    controllerSubscription?.dispose()
+    controller?.viewDisconnect()
+  }
 
-    this.runOnUIThread {
-      this.onBookOpenSucceededUI(controller)
-    }
+  private fun onBookOpenSucceeded(controller: SR2ControllerType) {
+    logger.debug("onBookOpenSucceeded")
+    runOnUIThread { onBookOpenSucceededUI(controller) }
   }
 
   @UiThread
   private fun onBookOpenSucceededUI(controller: SR2ControllerType) {
     this.controller = controller
-    controller.viewConnect(this.webView)
-    this.controllerHost.onControllerBecameAvailable(controller)
+    controller.viewConnect(webView)
+
+    controllerSubscription =
+      controller.events.subscribe { onControllerEvent(it) }
+
+    controllerHost.onControllerBecameAvailable(controller)
   }
 
-  private fun onBookOpenFailed(exception: Throwable) {
-    this.logger.error("onBookOpenFailed: ", exception)
+  private fun onBookOpenFailed(e: Throwable) {
+    logger.error("onBookOpenFailed: ", e)
+    runOnUIThread { onBookOpenFailedUI(e) }
+  }
 
-    this.runOnUIThread {
-      this.onBookOpenFailedUI(exception)
-    }
+  @UiThread
+  private fun onBookOpenFailedUI(e: Throwable) {
+    TODO()
   }
 
   @UiThread
-  private fun onBookOpenFailedUI(exception: Throwable) {
+  private fun onReadingPositionChanged(event: SR2Event.SR2ReadingPositionChanged) {
+    logger.debug("chapterTitle=${event.chapterTitle}")
+    progressView.apply { max = 100; progress = event.percent }
+    positionPageView.text = getString(R.string.progress_page, event.currentPage, event.pageCount)
+    positionTitleView.text = event.chapterTitle
+    positionPercentView.text = getString(R.string.progress_percent, event.percent)
+  }
 
+  private fun onControllerEvent(event: SR2Event) {
+    when (event) {
+      is SR2Event.SR2ReadingPositionChanged -> {
+        runOnUIThread { onReadingPositionChanged(event) }
+      }
+    }
   }
 
   private fun runOnUIThread(f: () -> Unit) {
     val looper = Looper.getMainLooper()
     val h = Handler(looper)
-    h.post {
-      f.invoke()
-    }
-  }
-
-  override fun onStop() {
```

Comment:
We don't disconnect the view on stopping anymore?

Reply:
We do! This method moved up to line 122-126. I'm used to grouping lifecycle methods together so it was moved up to be just after "onStop()".

Reply:
:+1: